clash-lib/src/Clash/Core/Evaluator/Models.hs

```diff
@@ -0,0 +1,258 @@
+-- as the 'AsTerm' class (which converts back to Term) has instances for both.
+--
+data Evaluator = Evaluator
+  { evaluate       :: Term -> State Env Value
+  , quote          :: Value -> State Env Nf
+  }
+
+-- Local bindings are not stored in a VarEnv, as we still want to be
+-- able to access the keys (VarEnv simply indexes by unique).
+
+type EnvTyMap   = Map TyVar Type
+type EnvTmMap   = Map Id (Either Term Value)
+
+-- EnvGlobals refers to global bindings in scope during evalaution (i.e. other
+-- top-level functions used in a program).
+
+type EnvGlobals = VarEnv (Binding (Either Term Value))
+
+-- EnvPrimsIO refers to the result of IO operations performed in primitives
+-- during evalaution e.g. creating a new ByteArray in the GHC frontend.
+
+type EnvPrimsIO = (IntMap Value, Int)
+
+-- | An Environment contains all in scope terms and types while evaluating.
+-- This consists of
+--
+--   * local types and terms from the term being evaluated
+--   * top-level definitions in scope (i.e. global bindings)
+--   * IO results of primitive operations
+--   * in scope uniques and a means of generating uniques
+--
+-- When running the partial evaluator at different times, it may be desirable
+-- to keep the globals (some of which may have been partially evaluated) and
+-- the results of primitives in IO for future runs.
+--
+data Env = Env
+  { envTypes   :: EnvTyMap
+  , envLocals  :: EnvTmMap
+  , envGlobals :: EnvGlobals
+  , envPrimsIO :: EnvPrimsIO
+  , envTcMap   :: TyConMap
+  , envInScope :: InScopeSet
+  , envSupply  :: Supply
+  }
+
+instance Show Env where
+  show env = show (envTypes env, envLocals env, envPrimsIO env)
+
+instance NFData Env where
+  rnf env = rnf (envTypes env)
+    `seq` rnf (envLocals env)
+    `seq` rnf (envGlobals env)
+    `seq` rnf (envPrimsIO env)
+
+mkEnv :: BindingMap -> EnvPrimsIO -> TyConMap -> InScopeSet -> Supply -> Env
+mkEnv bs = Env mempty mempty gs
+ where
+  gs = fmap Left <$> bs
+
+asSubst :: Env -> Subst
+asSubst env = subst2
+ where
+  subst0 = mkSubst (envInScope env)
+  subst1 = foldl' (\s -> uncurry $ extendTvSubst s) subst0 (Map.toList $ envTypes env)
+  subst2 = extendIdSubstList subst1 (Map.toList . fmap asTerm $ envLocals env)
+
+-- | Add a new term / value to the environment. The unique for the new
+-- element is guaranteed to be unique within the environment.
+--
+-- For new global bindings, we assume that it is possible to inline
+-- them later during evaluation.
+--
+insertEnv :: IdScope -> Id -> Either Term Value -> Env -> Env
+insertEnv scope i etv env =
+  case scope of
+    LocalId -> env
+      { envLocals = Map.insert i' etv (envLocals env)
+      , envInScope = extendInScopeSet (envInScope env) i'
+      }
+
+    GlobalId ->
+      let b = Binding i' noSrcSpan Inline etv
+       in env
+            { envGlobals = extendVarEnv i' b (envGlobals env)
+            , envInScope = extendInScopeSet (envInScope env) i'
```

Comment:
Global variable references are not free variables, but more akin to constants; i.e `\x[LocalId] -> x[GlobalId]` is the constant function, not identity. You therefor won't need to extend the `InScopeSet` for the regular reason: generating non-shadowing local identifiers.